pwnlib/elf/elf.py

```diff
@@ -611,40 +665,140 @@ def _populate_libraries(self):
+        dumps ``/proc/self/maps`` and exits, and **actually executing the binary**.
+
+        Returns:
+            A ``dict`` mapping file paths to the lowest address they appear at.
+            Does not do any translation for e.g. QEMU emulation, the raw results
+            are returned.
+
+            If there is not enough space to inject the shellcode in the segment
+            which contains the entry point, returns ``{}``.
+
+        Doctests:
+
+            These tests are just to ensure that our shellcode is correct.
+
+            >>> for arch in CAT_PROC_MAPS_EXIT:
+            ...   with context.local(arch=arch):
+            ...     sc = shellcraft.cat("/proc/self/maps")
+            ...     sc += shellcraft.exit()
+            ...     sc = asm(sc)
+            ...     sc = enhex(sc)
+            ...     assert sc == CAT_PROC_MAPS_EXIT[arch]
+        """
+
+        # Get our shellcode
+        sc = CAT_PROC_MAPS_EXIT.get(self.arch, None)
+
+        if sc is None:
+            log.error("Cannot patch /proc/self/maps shellcode into %r binary", self.arch)
+
+        sc = unhex(sc)
+
+        # Ensure there is enough room in the segment where the entry point resides
+        # in order to inject our shellcode.
+        seg = self.get_segment_for_address(self.entry, len(sc))
+        if not seg:
+            log.warn_once("Could not inject code to determine memory mapping for %r: Not enough space", self)
+            return {}
+
+        # Create our temporary file
+        # NOTE: We cannot use "with NamedTemporaryFile() as foo", because we cannot
+        # execute the file while the handle is open.
+        fd, path = tempfile.mkstemp()
+
+        # Close the file descriptor so that it may be executed
+        os.close(fd)
+
+        # Save off a copy of the ELF
+        self.save(path)
+
+        # Load a new copy of the ELF at the temporary file location
+        old = self.read(self.entry, len(sc))
         try:
-            cmd = 'ulimit -s unlimited; LD_TRACE_LOADED_OBJECTS=1 LD_WARN=1 LD_BIND_NOW=1 %s 2>/dev/null' % sh_string(self.path)
+            self.write(self.entry, sc)
+            self.save(path)
+        finally:
+            # Restore the original contents
+            self.write(self.entry, old)
 
-            data = subprocess.check_output(cmd, shell = True, stderr = subprocess.STDOUT)
-            libs = misc.parse_ldd_output(data)
+        # Make the file executable
+        os.chmod(path, 0o755)
 
-            for lib in dict(libs):
-                if os.path.exists(lib):
-                    continue
+        # Run a copy of it, get the maps
+        try:
+            with context.silent:
+                io = process(path)
+                data = io.recvall(timeout=2)
+        except Exception:
+            log.warn_once("Injected /proc/self/maps code did not execute correctly")
+            return {}
+
+        # Swap in the original ELF name
+        data = data.replace(path, elf.path)
```

Comment:
I think you meant `self.path`
```suggestion
        data = data.replace(path, self.path)
```